CDAP-13120 use correct directory for file in PartitionedFileSet

albertshau opened the pull request:
If there was no base path set for a PartitionedFileSet, it would
default to a 'files' directory whose parent is a directory named
after the dataset. This means when the dataset is deleted, only
the 'files' directory is deleted, but not the parent directory.
This behavior is especially problematic when using a
PartitionedFileSet as a local fileset in a workflow, as each run
of the workflow would leave a directory per PartitionedFileSet.

albertshau committed:
CDAP-13120 use correct directory for file in PartitionedFileSet

If there was no base path set for a PartitionedFileSet, it would
default to a 'files' directory whose parent is a directory named
after the dataset. This means when the dataset is deleted, only
the 'files' directory is deleted, but not the parent directory.
This behavior is especially problematic when using a
PartitionedFileSet as a local fileset in a workflow, as each run
of the workflow would leave a directory per PartitionedFileSet.

albertshau commented:
https://builds.cask.co/browse/CDAP-RUT1488

albertshau commented:
https://builds.cask.co/browse/CDAP-RUT1488

anew commented:
Agreed with the approach. However:
- I don't think this should be done in reconfigure(), unless it had been done in configure(). Otherwise,m when people reconfigure their existing PFS, it will change the base path, even though there may be already existing data in the old <name>/files base path.
- I don't see unit tests. It should be fairly straight-forward to write a test that creates the PFS, adds a partition, deletes the PFS and validates that all directories are gone.

albertshau committed:
CDAP-13120 make defaulting base path backwards compatible

Modify the reconfigure logic so that the base path is not
defaulted unless it was defaulted when it was created. Also adding
unit tests.

anew commented on cdap-data-fabric/src/test/java/co/cask/cdap/api/dataset/lib/SystemDatasetDefinitionTest.java:
```diff
@@ -227,6 +229,39 @@ public void testPFSReconfigure() {
+      .add(PartitionedFileSetDefinition.NAME_AS_BASE_PATH_DEFAULT, "false")
+      .build();
+    DatasetSpecification oldSpec = pfsDef.configure("pfs", oldProps);
+    try {
```
since this is a test, you could just add "throws IncompatibleUpdateException" to the method, and then you don't need to catch and wrap it here.

anew commented on cdap-data-fabric/src/test/java/co/cask/cdap/data2/datafabric/dataset/DatasetsUtilTest.java:
```diff
@@ -57,9 +58,14 @@ public void testFixProperties() throws DatasetManagementException, UnsupportedTy
-            FileSetProperties.builder().setBasePath("relative").build());
+            FileSetProperties.builder()
+              .setBasePath("relative")
+              .add(PartitionedFileSetDefinition.NAME_AS_BASE_PATH_DEFAULT, "true").build());
```
hmmm - because this sets a base path, the NAME_AS_BASE_PATH should not apply and not be needed, right?

albertshau replied:
ah you're right. This is leftover from when I had implemented compatibility incorrectly, will remove.

anew commented on cdap-data-fabric/src/test/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetTest.java:
```diff
@@ -211,6 +211,21 @@ public void testPermissions() throws Exception {
+    Assert.assertTrue(baseLocation.exists());
+    Assert.assertTrue(baseLocation.isDirectory());
+    dsFrameworkUtil.deleteInstance(id);
+    Assert.assertFalse(baseLocation.exists());
```
I think this would be true with the old behavior, too. If the base location is {name}/files, then that will be deleted...

To really test that it behaves correctly, I think you can create both a file set and pfs, get both base locations and assert that their parent is the same?

anew reviewed (commented):
just a couple of comments. 

It is amazing how much complexity the backward-compatibility adds to this code...

albertshau committed:
CDAP-13120 fix unit tests
